Basic pool request for DeFMO.

rozumden opened the pull request:
Inference implementation of DeFMO.

### Description
A few sentences describing the changes proposed in this pull request.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated


## PR Checklist
### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

ducha-aiki commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+
+import torch
+import torch.nn as nn
+import torchvision.models
```
That breaks basic kornia reliance on the torch-only. Could you please add (copy-paste) some definition of the relevant blocks to the defmo? I understand, that it is a bit more cumbersome, but being a dependence-less, is an improtant feature.

edgarriba replied:
+1 add whatever code network is proven that works well + proper reference from where it’s taken

rozumden replied:
Done.

ducha-aiki commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+        )
+        self.net = model
+
+    def forward(self, latent, times):
```
add type hints, similarly as for the encoder

ducha-aiki commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+        model = torchvision.models.resnet50(pretrained=True)
+        modelc1 = nn.Sequential(*list(model.children())[:3])
+        modelc2 = nn.Sequential(*list(model.children())[4:8])
+        pretrained_weights = modelc1[0].weight
```
Is the pretrained weight not stored in the encoder_best.pt?

rozumden replied:
It's initialization for "training". But you're right that it's not needed for the inference. Should I remove it?

rozumden replied:
Removed.

shijianjian commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+class EncoderDeFMO(nn.Module):
+    def __init__(self):
+        super(EncoderDeFMO, self).__init__()
+        model = torchvision.models.resnet50(pretrained=True)
```
One question here, shall we allow more models than resnet50? Probably adding those ```timm``` models avaliable as well?

ducha-aiki replied:
@shijianjian currently not, because kornia does not depends on timm. While we may need to find some solution for handling pretrained models with a standard backbones, I believe, it is not related to this particular PR

edgarriba replied:
The idea is to provide curated methods/models. /cc @ethanrublee @garybradski

If the paper states that resnet50 gives average best performance that’s what I would use. Later if a user wants to explore other options, since everything it’s exposed as a module, with python should be pretty straightforward to plug any model and fine tune. BTW, @rozumden do you have somewhere training code for this model ?

edgarriba replied:
great - sure we can define later a proper fine tuning api

ducha-aiki reviewed (changes requested):
Thank you for the great contribution! 

Could you also please add couple of the tests? You can adopt the ones from here
https://github.com/kornia/kornia/blob/master/test/feature/test_hardnet.py

at least, shape and gradcheck. 

cc @edgarriba - what do you think about the encoder stuff? I think for now, it is better, to add blocks definitions here, but later we have to deal with it in a more principled way.

One more thing - add the Module names to the documentation here 
https://github.com/kornia/kornia/blob/master/docs/source/feature.rst

And the citation to your paper here ;)
https://github.com/kornia/kornia/blob/master/docs/source/references.bib

shijianjian commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+        modelc1[0] = nn.Conv2d(6, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
+        modelc1[0].weight.data[:, :3, :, :] = nn.Parameter(pretrained_weights)
+        modelc1[0].weight.data[:, 3:, :, :] = nn.Parameter(pretrained_weights)
+        self.net = nn.Sequential(modelc1, modelc2)
```
I do not get it. Why split and merge again?

```python
pretrained_weights = model[0].weight
model[0] = nn.Conv2d(6, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
model[0].weight.data[:, :3, :, :] = nn.Parameter(pretrained_weights)
model[0].weight.data[:, 3:, :, :] = nn.Parameter(pretrained_weights)
```

Will this be not working?

edgarriba replied:
Agree

rozumden replied:
No, it's not the same because I merge [:3] and [4:8] parts, omitting one MaxPooling layer [4].

edgarriba replied:
gotcha :+1:

shijianjian commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+
+
+class DeFMO(nn.Module):
+    """
+    Module that disentangle a fast-moving object from the background and performs deblurring.
+    This is based on the original code from paper "DeFMO: Deblurring and Shape Recovery
+        of Fast Moving Objects". See :cite:`DeFMO2021` for more details.
+    Args:
+        pretrained: (bool) Download and set pretrained weights to the model. Default: false.
+    Returns:
+        torch.Tensor: Temporal super-resolution without background.
+    Shape:
+        - Input: (B, 6, H, W)
+        - Output: (B, S, 4, H, W)
+    Examples:
+        >>> input = torch.rand(16, 6, 240, 320)
+        >>> defmo = DeFMO()
+        >>> tsr_nobgr = defmo(input) # 16x24x4x240x320
+    """
```
Add new lines between sections. arg typing can be generated automatically now. You may remove it.

shijianjian reviewed (changes requested):
LGTM. Just some dummy questions. Please refine docs and add tests.

Also, are you willing to add a minimal example to demonstrate the ability of this module in [tutorials](https://kornia-tutorials.readthedocs.io/en/latest/)?

edgarriba commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+            result = self.net(latenti)
+            renders.append(result)
+        renders = torch.stack(renders, 1).contiguous()
+        renders[:, :, :4] = nn.Sigmoid()(renders[:, :, :4])
```
Be careful with In place stuff since will raise some backprop issues

edgarriba commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+
+class DeFMO(nn.Module):
+    """
+    Module that disentangle a fast-moving object from the background and performs deblurring.
```
```suggestion
    Module that disentangle a fast-moving object from the background and performs deblurring.

```

edgarriba commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+            renders[ki, shuffled_times[ki, :]] = renders[ki, :].clone()
+        return renders
+
+
```
It’s fine for now, but thinking in future - should we place this in filters or somewhere where we have blurring ? 

Besides, I believe we could have a more generic API for `DeBlurring(…)` and the network just being the selected/curated backbone method. /cc @ethanrublee @garybradski

rozumden commented on kornia/feature/defmo.py:
```diff
@@ -0,0 +1,112 @@
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        latent = self.encoder(input)
+        times = torch.linspace(0, 1, self.tsr_steps).to(input.device)[None].repeat(input.shape[0], 1)
```
Done.

rozumden commented:
Thanks for your comments. I've added most of your changes and updated the mentioned citations, etc. I've put an example in 'examples/defmo_example.py'.

edgarriba reviewed (commented):
@rozumden the example file - should be a notebook a placed in the following repo: https://github.com/kornia/tutorials and the files should go here: https://github.com/kornia/data

Let's do the following. We merge first the implementation and then we think about the example/tutorial. After taking a quick look I believe that some things can be kornialized.

rozumden commented:
Hi, I agree that we can merge the implementation first. What else is missing or you want me to fix?

edgarriba commented on examples/defmo_example.py:
```diff
@@ -0,0 +1,149 @@
+import cv2
```
@rozumden skip this example for now from this PR

ducha-aiki replied:
I agree

rozumden replied:
Done.

edgarriba reviewed (dismissed):
for me just remove the example and it's fine. @ducha-aiki

edgarriba commented on test/feature/test_defmo.py:
```diff
@@ -0,0 +1,37 @@
+
+
+class TestDeFMO:
+    def test_shape(self, device):
```
```suggestion
    def test_shape(self, device, dtype):
```

edgarriba commented on test/feature/test_defmo.py:
```diff
@@ -0,0 +1,37 @@
+
+class TestDeFMO:
+    def test_shape(self, device):
+        inp = torch.ones(1, 6, 240, 320, device=device)
```
```suggestion
        inp = torch.ones(1, 6, 240, 320, device=device, dtype=dtype)
```

edgarriba commented on test/feature/test_defmo.py:
```diff
@@ -0,0 +1,37 @@
+class TestDeFMO:
+    def test_shape(self, device):
+        inp = torch.ones(1, 6, 240, 320, device=device)
+        defmo = DeFMO().to(device)
```
```suggestion
        defmo = DeFMO().to(device, dtype)
```

edgarriba reviewed (commented):
add `dtype` in the tests. I see somehow the module is not well imported. You need to define here: https://github.com/kornia/kornia/blob/master/kornia/feature/__init__.py

ducha-aiki commented on test/feature/test_defmo.py:
```diff
@@ -9,31 +9,31 @@
     @staticmethod
     @pytest.mark.skip("jacobian not well computed")
-    def test_gradcheck(self, device):
+    def test_gradcheck(selfarg, device):
```
Replace
>def test_gradcheck(selfarg, device): 
with 
def test_gradcheck(self, device, dtype):

everywhere. That is why tests are failing

edgarriba commented:
I see that you also forgot to remove the image files from the example

edgarriba commented on kornia/__init__.py:
```diff
@@ -51,6 +51,7 @@
     zca_whiten,
 )
 from kornia.feature import gftt_response, harris_response, hessian_response, MKDDescriptor, nms2d, SIFTDescriptor
+from kornia.feature import DeFMO
```
@ducha-aiki I'd keep it under `kornia.feature` instead of top level. Probably we need to review all the namespaces organization at some point.

ducha-aiki replied:
Well, it does not really fit into the (local) feature module, tbh and because of being high-level application, I think would benefit from being exposed. 
And yes, we need to review and define some policy

edgarriba commented:
@rozumden @ducha-aiki loking forward for a demo :))

edgarriba commented:
@ducha-aiki @rozumden would be possible to have a simple notebook showing the usage of this API ? would like to make a release by end of the week and have some material to promote this new functionality

ducha-aiki commented:
That's where now deleted example https://github.com/rozumden/kornia/commit/025ef024c6ccdc22714e6c04a873f55588ff703a will come handy :)
